src/widgets/list.rs

```diff
@@ -1598,6 +1642,48 @@ mod tests {
         assert_buffer_eq!(buffer, expected);
     }
 
+    #[test]
+    fn test_offset_without_selecting() {
+        let items = list_items(vec![
+            "Item 0", "Item 1", "Item 2", "Item 3", "Item 4", "Item 5",
+        ]);
+        let list = List::new(items);
+        let mut state = ListState::default();
+
+        // Select an item
+        state.select(Some(4));
+
+        // Render the item, which modifies the offset of the state
+        let buffer = render_stateful_widget(list, &mut state, 10, 3);
+        let expected = Buffer::with_lines(vec!["Item 2    ", "Item 3    ", "Item 4    "]);
+        assert_buffer_eq!(buffer, expected);
+
+        // Now unselect the item, but keep its position
+        let offset = state.offset;
+        state.select(None);
+        *state.offset_mut() = offset;
+
+        // Re-render the list, unselected but with an offset
+        let items = list_items(vec![
+            "Item 0", "Item 1", "Item 2", "Item 3", "Item 4", "Item 5",
+        ]);
+        let list = List::new(items);
+        let buffer = render_stateful_widget(list, &mut state, 10, 3);
+        let expected = Buffer::with_lines(vec!["Item 2    ", "Item 3    ", "Item 4    "]);
+        assert_buffer_eq!(buffer, expected);
+
+        // Re-render the list, unselected but with an offset that will leave empty spaces
+        // in the list
+        let items = list_items(vec![
+            "Item 0", "Item 1", "Item 2", "Item 3", "Item 4", "Item 5",
+        ]);
+        *state.offset_mut() = 4;
+        let list = List::new(items);
+        let buffer = render_stateful_widget(list, &mut state, 10, 3);
+        let expected = Buffer::with_lines(vec!["Item 4    ", "Item 5    ", "          "]);
+        assert_buffer_eq!(buffer, expected);
+    }
```

Comment:
IMO, this does not clearly communicate the intent of your change and the fixed behavior. The underlying problem (that you fixed) is the offset not being taken into account while rendering.
I think testing the correct render of a list with a given offset is enough.

Reply:
I half-agree, with the reasoning being because rendering a List manipulates the list's state, taking the offset change into account. Each `assert_eq` here verifies each step of list manipulation with offset modifications either through the List's code or by user intervention. Since `List::get_item_bounds()` uses this offset change and determines a new offset, it seemed fair/correct to include each of these tests.

Previously Josh mentioned why one test was failing - this unit test covers that scenario with the last `assert_eq` and is expected behavior for setting an offset and rendering it, without having rendered the list before (since the List's render code changes the offset as mentioned earlier).

Happy to remove one or more of these tests if you still consider it unnecessary, but I still lean towards each of these tests being relevant. Perhaps moving one of those `assert_eq`s would make things clearer.

Reply:
I agree that this test is useful, but I think the underlying thing you fixed is just the offset not being taken into account (selection or not). Previously setting `offset` explicitly would do nothing, I think this is the testcase we're missing (your testcase covers it but too broadly, it's not its main focus). Hope you get what I mean here.

Overall, I'd propose moving this to an integration test (it seems more like testing a usage rather than one particular feature) and testing that the `offset` setter works as expected (i.e. renders a shifted list).

Reply:
```rust
#[test]
fn test_offset_renders_shifted(
) {
    let items = list_items(vec![
        "Item 0", "Item 1", "Item 2", "Item 3", "Item 4", "Item 5", "Item 6",
    ]);
    let list = List::new(items);
    let mut state = ListState::default().with_offset(3);
    let buffer = render_stateful_widget(list, &mut state, 6, 3);

    let expected = Buffer::with_lines(vec!["Item 3", "Item 4", "Item 5"]);
    assert_buffer_eq!(buffer, expected);
}
```

Setting an offset didn't work previously and is now fixed with your changes. This test demonstrate that.

Reply:
I understand what you're saying, and agree this is an appropriate test.

Updated and pushed the test to reflect this. Thanks!